napari/config/_tests/test_config.py

```diff
@@ -0,0 +1,549 @@
+    with config.set(abc=123):
+        assert config.config["abc"] == 123
+        with config.set(abc=456):
+            assert config.config["abc"] == 456
+        assert config.config["abc"] == 123
+
+    assert "abc" not in config.config
+
+    with config.set({"abc": 123}):
+        assert config.config["abc"] == 123
+    assert "abc" not in config.config
+
+    with config.set({"abc.x": 1, "abc.y": 2, "abc.z.a": 3}):
+        assert config.config["abc"] == {"x": 1, "y": 2, "z": {"a": 3}}
+    assert "abc" not in config.config
+
+    d = {}
+    config.set({"abc.x": 123}, config=d)
+    assert d["abc"]["x"] == 123
+
+
+def test_set_kwargs():
+    with config.set(foo__bar=1, foo__baz=2):
+        assert config.config["foo"] == {"bar": 1, "baz": 2}
+    assert "foo" not in config.config
+
+    # Mix kwargs and dict, kwargs override
+    with config.set({"foo.bar": 1, "foo.baz": 2}, foo__buzz=3, foo__bar=4):
+        assert config.config["foo"] == {"bar": 4, "baz": 2, "buzz": 3}
+    assert "foo" not in config.config
+
+    # Mix kwargs and nested dict, kwargs override
+    with config.set({"foo": {"bar": 1, "baz": 2}}, foo__buzz=3, foo__bar=4):
+        assert config.config["foo"] == {"bar": 4, "baz": 2, "buzz": 3}
+    assert "foo" not in config.config
+
+
+def test_set_nested():
+    with config.set({"abc": {"x": 123}}):
+        assert config.config["abc"] == {"x": 123}
+        with config.set({"abc.y": 456}):
+            assert config.config["abc"] == {"x": 123, "y": 456}
+        assert config.config["abc"] == {"x": 123}
+    assert "abc" not in config.config
+
+
+def test_set_hard_to_copyables():
+    import threading
+
+    with config.set(x=threading.Lock()):
+        with config.set(y=1):
+            pass
+
+
+def test_set_raises():
+    # a single positional arg should be a dict...
+    with pytest.raises(TypeError):
+        config.set('asdf.sadf')
+
+
+@pytest.mark.parametrize("mkdir", [True, False])
+def test_ensure_file_directory(mkdir, tmpdir):
+    a = {"x": 1, "y": {"a": 1}}
+
+    source = os.path.join(str(tmpdir), "source.yaml")
+    dest = os.path.join(str(tmpdir), "dest")
+
+    with open(source, "w") as f:
+        yaml.dump(a, f)
+
+    if mkdir:
+        os.mkdir(dest)
+
+    config.ensure_file(source=source, destination=dest)
+
+    assert os.path.isdir(dest)
+    assert os.path.exists(os.path.join(dest, "source.yaml"))
+
+
+def test_ensure_file_defaults_to_NAPARI_CONFIG_directory(tmpdir):
+    a = {"x": 1, "y": {"a": 1}}
+    source = os.path.join(str(tmpdir), "source.yaml")
+    with open(source, "w") as f:
+        yaml.dump(a, f)
+
+    destination = os.path.join(str(tmpdir), "napari")
+    PATH = config.core.PATH
+    try:
+        config.core.PATH = destination
+        config.ensure_file(source=source)
+    finally:
+        config.core.PATH = PATH
+
+    assert os.path.isdir(destination)
+    [fn] = os.listdir(destination)
+    assert os.path.split(fn)[1] == os.path.split(source)[1]
+
+
+def test_rename():
```

Comment:
If I have the following config in the console
```python
config.config
Out[36]: {'perfmon': 1, 'main_window': {'theme': 'dark'}}
```
and I do
```python
config.rename({'main_window': 'new_window'})
config.config
Out[38]: {'perfmon': 1, 'new_window': {'theme': 'dark'}}
```
that works as expected, but if I then try and rename `theme` I get the following
```python-traceback
config.rename({'new_window.theme': 'new_window.color'})
---------------------------------------------------------------------------
KeyError                                  Traceback (most recent call last)
<ipython-input-39-cc1db23ba1be> in <module>
----> 1 config.rename({'new_window.theme': 'new_window.color'})

~/GitHub/napari/napari/config/core.py in rename(aliases, config)
    507 
    508     for k in old:
--> 509         del config[canonical_name(k, config)]  # TODO: support nested keys
    510 
    511     _set(new, config=config)

KeyError: 'new_window.theme'
```
Should that have worked? Or I am doing something wrong?

Reply:
nested keys aren't supported for `rename` yet ... (it's in the stack trace there: `# TODO: support nested keys`).  I don't think rename will be used all that often though